x-pack/plugins/ml/public/application/services/analytics_fields_service.ts

```diff
@@ -0,0 +1,150 @@
+export function loadAnalyticsFields(
+  indexPatternId: string,
+  savedSearchId: string,
+  indexPatterns: IndexPatternsContract
+) {
+  return new Promise(async (resolve, reject) => {
+    if (indexPatternId !== undefined) {
+      // index pattern is being used
+      const indexPattern: IIndexPattern = await indexPatterns.get(indexPatternId);
+      await analyticsFieldsService.initializeFromIndexPattern(indexPattern);
+      resolve(analyticsFieldsService.fieldsAndAggs);
+    } else if (savedSearchId !== undefined) {
+      // saved search is being used
+      // load the index pattern from the saved search
+      const { indexPattern } = await getIndexPatternAndSavedSearch(savedSearchId);
+      if (indexPattern === null) {
+        // eslint-disable-next-line no-console
+        console.error('Cannot retrieve index pattern from saved search');
+        reject();
+        return;
+      }
+      await analyticsFieldsService.initializeFromIndexPattern(indexPattern);
+      resolve(analyticsFieldsService.fieldsAndAggs);
+    } else {
+      reject();
+    }
+  });
+}
+
+// create two lists, one removing text fields if there are keyword equivalents and vice versa
+function processTextAndKeywordFields(fields: Field[]) {
+  const keywordIds = fields.filter((f) => f.type === ES_FIELD_TYPES.KEYWORD).map((f) => f.id);
+  const textIds = fields.filter((f) => f.type === ES_FIELD_TYPES.TEXT).map((f) => f.id);
+
+  const fieldsPreferringKeyword = fields.filter(
+    (f) =>
+      f.type !== ES_FIELD_TYPES.TEXT ||
+      (f.type === ES_FIELD_TYPES.TEXT && keywordIds.includes(`${f.id}.keyword`) === false)
+  );
+
+  const fieldsPreferringText = fields.filter(
+    (f) =>
+      f.type !== ES_FIELD_TYPES.KEYWORD ||
+      (f.type === ES_FIELD_TYPES.KEYWORD &&
+        textIds.includes(f.id.replace(/\.keyword$/, '')) === false)
+  );
+
+  return { fieldsPreferringKeyword, fieldsPreferringText };
+}
+
+// Keep top nested field and remove all <nested_field>.* fields
+function removeNestedFieldChildren(resp: any, indexPatternTitle: string) {
+  const results = resp[indexPatternTitle];
+  const fields: Field[] = [];
+  const nestedFields: Record<string, boolean> = {};
+
+  if (results !== undefined) {
+    results.fields.forEach((field: Field) => {
+      if (field.type === ES_FIELD_TYPES.NESTED && nestedFields[field.name] === undefined) {
+        nestedFields[field.name] = true;
+        fields.push(field);
+      }
+    });
+
+    if (Object.keys(nestedFields).length > 0) {
+      results.fields.forEach((field: Field) => {
+        if (field.type !== ES_FIELD_TYPES.NESTED) {
+          const fieldNameParts = field.name.split('.');
+          const rootOfField = fieldNameParts.shift();
+          if (rootOfField && nestedFields[rootOfField] === undefined) {
+            fields.push(field);
+          }
+        }
+      });
+    } else {
+      fields.push(...results.fields);
+    }
+  }
+  return fields;
+}
+
+class AnalyticsFieldsService {
```

Comment:
seeing as this is so similar to `NewJobCapsService` I think they should have a common name and a probably a common base class.
`NewJobCapsService` isn't the best name, so i'm fine with renaming it.
The same goes for `loadAnalyticsFields` above, it appears to be identical to the one in `new_job_capabilities_service.ts`, so a common function could be created which takes a job type.

Reply:
Yeah I thought of that initially but was leaning toward having distinct classes since the use cases have been diverging. With the work that will be going in for the workspace/compare view in DFA and other things we'll be adding I thought we might start to diverge on needs even more. 🤔 

You do have a point with the `loadAnalyticsFields` function as those are identical but I'm not sure having a common base class would give us any benefits - the `initializeFromIndexPattern`, for example, is more bare-bones.

I suppose we could have a super basic base class. 🤔

Reply:
After chatting with @jgowdyelastic - agreed that we should have a base class and shared utils in a separate directory which both classes can then use.

Reply:
Created a common base class and moved shared functionality in 0e52a561d3f69be116ab4b5a165dbe9952d1d080